taskvine/src/tools/vine_plot_txn_log

```diff
@@ -687,31 +710,55 @@ class TxnPlotWorkers(TxnPlot):
+        cache_updates = worker_xfers[ worker_xfers["direction"] == "CACHE_UPDATE" ]
+        slot_height = 1
+
+        # plot the red bar: cache updates
+        self.bh(fig, base_slot, cache_updates["start_time"], cache_updates["wall_time"], 
+                self.legend_dask["data transfers"], height=slot_height, type="cu_dask")
+
+        # every following task's bar slot add 1
+        worker_tasks["slot"] += 1
+
+        # plot the green bar: tasks executing
+        self.bh(fig, retrieved_tasks["slot"], retrieved_tasks["time_worker_start"], 
+                retrieved_tasks["time_worker_end"] - retrieved_tasks["time_worker_start"], 
+                self.legend_dask["tasks executing"], height=slot_height, type="te_dask")
+        
+        # determine the following starting slot for tasks in the next worker
+        return worker_tasks["slot"].max() + 0.2
 
     def plot(self, fig, manager):
-        ts = manager.tasks
-        xs = manager.transfers
-        ws = manager.workers
-
-        gs = ts.groupby("worker_id", sort=False)
-        by_min_connects = ws.groupby("worker_id")["CONNECTION"].min().sort_values().index
-        gsd = {wid: g  for (wid, g) in gs}
-
-        base_slot=0
-        for wid in by_min_connects:
-            worker = gsd.get(wid, None)
-            worker_times = ws[ws["worker_id"] == wid]
-            worker_xfers = xs[xs["worker_id"] == wid]
-
-            if worker is not None:
-                base_slot = self.plot_by_worker(fig, worker, worker_times, worker_xfers, base_slot)
+        tasks = manager.tasks
+        transfers = manager.transfers
+        workers = manager.workers
+
+        grouped_tasks = tasks.groupby("worker_id", sort=False)
+        grouped_tasks_dict = {worker_id: group  for (worker_id, group) in grouped_tasks}
+        grouped_workers = workers.groupby("worker_id")["CONNECTION"].min().sort_values().index
+        base_slot = 0
+        t=0
+        for worker_id in grouped_workers:
+            current_worker_tasks = grouped_tasks_dict.get(worker_id, None)
+            current_worker_times = workers[workers["worker_id"] == worker_id]
+            current_worker_transfers = transfers[transfers["worker_id"] == worker_id]
+            
+            if current_worker_tasks is not None:
+                if args.dask_colors:
+                    base_slot = self.plot_by_worker_dask(fig, current_worker_tasks, current_worker_times, current_worker_transfers, base_slot)
+                else:
+                    base_slot = self.plot_by_worker(fig, current_worker_tasks, current_worker_times, current_worker_transfers, base_slot)
 
         fig.set_ylabel("workers")
         fig.set_yticks((0, base_slot))
-        fig.set_yticklabels(["0", str(gs.ngroups)])  # ytick number of workers
+        fig.set_yticklabels(["0", str(grouped_tasks.ngroups)])  # ytick number of workers
 
-        if not self.opts.expand_waiting:
+        if not args.dask_colors and not self.opts.expand_waiting:
```

Comment:
Since you are using legend_dask directly in plot_dask, you don't need to add dask_colors here.